File: ios/Sources/Mekuri/MekuriLeafShadow.swift
import SwiftUI

/// The shadows a two-sided leaf casts: the contact shadow past the rim and
/// the crease ramp short of the crease. Drawn in a layer of the leaf's own
/// size beneath the leaf and over the revealed page, with the leaf's
/// progress, direction and configuration. Nothing is drawn at rest.
struct MekuriLeafShadow: View {
    private let progress: CGFloat
    private let direction: MekuriDirection
    private let configuration: MekuriConfiguration

    init(progress: CGFloat, direction: MekuriDirection, configuration: MekuriConfiguration) {
        self.progress = progress
        self.direction = direction
        self.configuration = configuration
    }

    var isFolded: Bool { self.progress > 0 }

    var foldPass: MekuriFoldPass {
        MekuriFoldPass(direction: self.direction, progress: self.progress)
    }

    var body: some View {
        GeometryReader { proxy in
            if self.isFolded {
                self.shadowLayer(size: proxy.size)
            }
        }
    }

    /// The layer is opaque so it is never culled; the shader replaces every
    /// pixel of it and samples none.
    private func shadowLayer(size: CGSize) -> some View {
        let pass = self.foldPass
        return Color.black
            .compositingGroup()
            .layerEffect(self.shadow(size: size, pass: pass), maxSampleOffset: .zero)
            .scaleEffect(x: pass.mirrorScale, y: 1)
    }

    private func shadow(size: CGSize, pass: MekuriFoldPass) -> Shader {
        MekuriFoldShader.make(
            size: size,
            progress: pass.shaderProgress,
            configuration: self.configuration,
            face: .shadow,
            hinge: MekuriFoldShader.hinge(in: size)
        )
    }
}

File: android/mekuri/src/main/kotlin/studio/anicanon/mekuri/MekuriLeafShadow.kt
package studio.anicanon.mekuri

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.graphics.Color

/**
 * The shadows a two-sided leaf casts: the contact shadow past the rim and the
 * crease ramp short of the crease. Drawn in a layer of the leaf's own size
 * beneath the leaf and over the revealed page, so each shadow composites once
 * and the leaf's own faces carry none. Nothing is drawn at rest.
 */
@Composable
internal fun MekuriLeafShadow(
    progress: () -> Float,
    direction: MekuriDirection,
    configuration: MekuriConfiguration,
    modifier: Modifier = Modifier,
) {
    val isFolded by remember(progress) { derivedStateOf { progress() > 0f } }
    val shader = rememberMekuriFoldShader()
    Box(modifier) {
        if (!isFolded) return@Box
        Box(
            Modifier
                .fillMaxSize()
                .mekuriFold(
                    shader = shader,
                    mirrorScale = MekuriFoldPass.mirrorScale(direction),
                    face = MekuriFace.Shadow,
                    isHinged = true,
                    configuration = configuration,
                    progress = { MekuriFoldPass(direction, progress()).shaderProgress },
                )
                // The layer is opaque so it is never culled; the shader replaces
                // every pixel of it and samples none.
                .drawBehind { drawRect(Color.Black) },
        )
    }
}
